pennylane/ops/qubit/observables.py

```diff
@@ -159,9 +176,6 @@ def _matrix(cls, *params):
             raise TypeError("Observable must be a scipy sparse coo_matrix.")
         return A
 
-    def diagonalizing_gates(self):
-        return []
```

Comment:
I am not sure what the intention of this is, but it seems misleading to pretend the Hamiltonian is diagonal?

Reply:
This must have been to avoid an error that was occurring downstream due to something assuming `diagonalizing_gates` exists. Might be worth checking the PR that introduces `SparseHamiltonian`, maybe #1579?

Reply:
that was the wrong PR link 😆

Reply:
@soranjh I wonder if you could have a look...do you remember if this method has any significance? 

Or can we remove it (since it implies that all `SparseHamiltonian` objects are diagonal)?

Reply:
We added this for consistency with other observables that do not have specific `diagonalizing_gates`. We also have the same method for the [Projector](https://github.com/PennyLaneAI/pennylane/blob/c6021b3f7d544f58313dbc2534914566e2b52301/pennylane/ops/qubit/observables.py#L250) class. 

I removed this method in a branch and performed a VQE task with a sparse Hamiltonian for BeH2 and the results were accurate. Not sure if removing the method affects other tests or not.

Reply:
Thanks @soranjh!

I think even before this PR, "no diag gates defined" was different from "empty diaggates", so I hope we do ourselves a favour by removing this here - it suggests that the op is diagonal... (and a sign how hard it is for developers to see through the old system of representations).

Or @josh146 do the generators give us a problem? I fear even in this case we need to try and return the correct gates here, so I remove it for now!

Reply:
@mariaschuld now that I am fresh from doing the generator refactor, so much of the codebase assumes that:

- `op.diagonalizing_gates()` raises not implemented error: not defined
- `op.diagonalizing_gates()` returns empty list: diagonal!

which is quite nice? But yes I agree, we should reserve an empty list for diagonal operators (e.g., `projector`) and instead use `None` (or a registration system) for those that don't define diagonalizing gates.